File: python/src/square_root.py
# Heron's formula
def heron(number, estimate):
    return (number / estimate + estimate) / 2

# absolute value
def abs(number):
    if number < 0:
        return -number
    return number

# calculate square root using Heron's formula
def square_root(number):
    s0 = number
    s1 = number / 3
    
    while abs(s1 - s0) > 0.00001:
        s2 = heron (number, s1)
        s0 = s1
        s1 = s2
    return s1

File: java/src/main/java/SquareRoot.java
public class SquareRoot {
    public int squareRoot(final int radicand) {
        if (radicand < 0) {
            throw new IllegalArgumentException("Radicand must be non-negative.");
        }
        if (radicand <= 1) {
            return radicand;
        }
        long below = 0;
        long above = radicand;
        long attempt;
        long diff;

        do {
            attempt = (above + below) / 2;
            diff = attempt * attempt - radicand;
            if (diff > 0) {
                above = attempt;
            } else if (diff < 0) {
                if (attempt == below) {
                    throw  new IllegalArgumentException("Radicand does not have an integer square root.");
                }
                below = attempt;
            }
        } while (diff != 0);

        return (int)attempt;
    }
}
